cmd/aio/main.go

```diff
@@ -0,0 +1,134 @@
+package main
+
+import (
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+
+	proto "github.com/container-storage-interface/spec/lib/go/csi"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/hetznercloud/csi-driver/api"
+	"github.com/hetznercloud/csi-driver/app"
+	"github.com/hetznercloud/csi-driver/driver"
+	"github.com/hetznercloud/csi-driver/volumes"
+	"github.com/hetznercloud/hcloud-go/hcloud/metadata"
+)
+
+var logger log.Logger
+
+func main() {
+	logger = app.CreateLogger()
+
+	m := app.CreateMetrics(logger)
+
+	hcloudClient, err := app.CreateHcloudClient(m.Registry(), logger)
+	if err != nil {
+		level.Error(logger).Log(
+			"msg", "failed to initialize hcloud client",
+			"err", err,
+		)
+		os.Exit(1)
+	}
+
+	metadataClient := metadata.NewClient(metadata.WithInstrumentation(m.Registry()))
+
+	server, err := app.GetServer(logger, hcloudClient, metadataClient)
+	if err != nil {
+		level.Error(logger).Log(
+			"msg", "failed to fetch server",
+			"err", err,
+		)
+		os.Exit(1)
+	}
+
+	// node
+	serverID, err := metadataClient.InstanceID()
+	if err != nil {
+		level.Error(logger).Log("msg", "failed to fetch server ID from metadata service", "err", err)
+		os.Exit(1)
+	}
+
+	serverAZ, err := metadataClient.AvailabilityZone()
+	if err != nil {
+		level.Error(logger).Log("msg", "failed to fetch server availability-zone from metadata service", "err", err)
+		os.Exit(1)
+	}
+	parts := strings.Split(serverAZ, "-")
+	if len(parts) != 2 {
+		level.Error(logger).Log("msg", fmt.Sprintf("unexpected server availability zone: %s", serverAZ), "err", err)
+		os.Exit(1)
+	}
+	serverLocation := parts[0]
+
+	level.Info(logger).Log("msg", "Fetched data from metadata service", "id", serverID, "location", serverLocation)
```

Comment:
Lines 37-44 are from `cmd/controller` while 46-65 are from `cmd/node`. Both of these retrieve the server metadata.

You can remove one of these and just use the return value of the other to replace any usages.

E.g. you could replace

- `server.Datacenter.Location.Name` -> `serverLocation`

Or the other way round:

- `serverID` -> `server.ID`
- `serverLocation` -> `server.Datacenter.Location.Name`

## Background

We have these two differing implementations as the both serve different niches:

- The controller implementation should not rely on the metadata service, as the controller part may run on non-hcloud servers
- The node implementation may not use the Hetzner Cloud api, as access to it is not necessary and we do not need to distribute the API access token to all nodes this way

As this binary needs to access the hcloud API AND the metadata service anyway, it does not really matter which one you choose.